src/LaunchDarkly.EventSource/EventSource.cs

```diff
@@ -138,28 +144,70 @@ public async Task StartAsync()
+                }
+
+                SetReadyState(ReadyState.Connecting);
+                cancellationToken = newRequestTokenSource.Token;
+
                 try
                 {
-                    CancellationTokenSource newRequestTokenSource = null;
-                    lock (this)
-                    {
-                        if (_readyState == ReadyState.Shutdown)
-                        {
-                            // in case Close() was called in between the previous ReadyState check and the creation of the new token
-                            return;
-                        }
-                        newRequestTokenSource = new CancellationTokenSource();
-                        _currentRequestToken?.Dispose();
-                        _currentRequestToken = newRequestTokenSource;
-                    }
-                    await ConnectToEventSourceAsync(newRequestTokenSource.Token);
+                    await ConnectToEventSourceAsync(cancellationToken);
+
+                    // ConnectToEventSourceAsync normally doesn't return, unless it detects that the request has been cancelled.
+                    Close(ReadyState.Closed);
                 }
                 catch (Exception e)
-                {
-                    _logger.ErrorFormat("Encountered an error connecting to EventSource: {0}", e, e.Message);
-                    _logger.Debug("", e);
-                }
+                {
+                    CancelCurrentRequest();
+
+                    // If the user called Close(), ReadyState = Shutdown, so errors are irrelevant.
+                    if (ReadyState != ReadyState.Shutdown)
+                    {
+                        Close(ReadyState.Closed);
+
+                        Exception realException = e;
+                        if (e is OperationCanceledException oe)
+                        {
+                            // This exception could either be the result of us explicitly cancelling a request, in which case we don't
+                            // need to do anything else, or it could be that the request timed out.
+                            if (oe.CancellationToken.IsCancellationRequested)
+                            {
+                                realException = null;
+                            }
+                            else
+                            {
+                                realException = new TimeoutException();
+                            }
+                        }
+
+                        if (realException != null)
+                        {
+                            _logger.ErrorFormat("Encountered an error connecting to EventSource: {0}", realException, realException.Message);
+                            _logger.Debug("", realException);
+
+                            OnError(new ExceptionEventArgs(realException));
+                        }
+                    }

```

Comment:
This b lock is just a consolidation of the error-handling logic with code that used to be in `ConnectToEventSourceAsync`. We were handling some errors down there and rethrowing others up here, for no particular reason. I also added the logic on lines 188-200 that detects timeouts (which in .NET are not actually thrown as `TimeoutException` but as `OperationCanceledException` for some reason) and avoids showing stacktraces for pointless exceptions that we deliberately caused.